test/parallel_tests/test_bas.py

```diff
@@ -0,0 +1,232 @@
+        self._syn_delay = syn_delay
+        self._create_cells(r)
+        self._connect_cells()
+        ### have one netstim on each rank
+        if pc.gid_exists(pc.id()):
+            self._netstim = h.NetStim()
+            self._netstim.number = 1
+            self._netstim.start = stim_t
+            self._nc = h.NetCon(
+                self._netstim, pc.gid2cell(pc.id()).syn
+            )  ### grab cell with gid==pc.id() wherever it exists
+            self._nc.delay = stim_delay
+            self._nc.weight[0] = stim_w
+
+    def set_gids(self):
+        """Set the gidlist on this host."""
+        #### Round-robin counting.
+        #### Each host has an id from 0 to pc.nhost() - 1.
+        self.gidlist = list(range(pc.id(), self._N, pc.nhost()))
+        for gid in self.gidlist:
+            pc.set_gid2node(gid, pc.id())
+
+    def _create_cells(self, r):
+        self.cells = []
+        for i in self.gidlist:  ### only create the cells that exist on this host
+            theta = i * 2 * h.PI / self._N
+            self.cells.append(
+                BallAndStick(i, h.cos(theta) * r, h.sin(theta) * r, 0, theta)
+            )
+        ### associate the cell with this host and gid
+        for cell in self.cells:
+            pc.cell(cell._gid, cell._spike_detector)
+
+    def _connect_cells(self):
+        for target in self.cells:
+            source_gid = (target._gid - 1 + self._N) % self._N
+            nc = pc.gid_connect(source_gid, target.syn)
+            nc.weight[0] = self._syn_w
+            nc.delay = self._syn_delay
+            target._ncs.append(nc)
+
+
+def prun(tstop, restore=False):
+    pc.set_maxstep(10 * ms)
+
+    if restore:
+        assert np.allclose(h.t, tstop / 2)
+        ns = h.SaveState()
+        sf = h.File("state%d.bin" % pc.id())
+        ns.fread(sf)
+        ns.restore(1)
+        sf.close()
+    else:
+        assert np.allclose(h.t, 0)
+        pc.psolve(tstop / 2)
+        ss = h.SaveState()
+        ss.save()
+
+        sf = h.File("state%d.bin" % pc.id())
+        ss.fwrite(sf)
+        sf.close()
+    pc.psolve(tstop)
+
+
+def get_all_spikes(ring):
+    local_data = {cell._gid: list(cell.spike_times) for cell in ring.cells}
+    all_data = pc.py_allgather([local_data])
+
+    pc.barrier()
+    pc.done()
+
+    data = {}
+    for d in all_data:
+        data.update(d[0])
+
+    return data
+
+
+def compare_dicts(dict1, dict2):
+    keylist = dict1.keys()
+    array_1 = np.array([dict1[key] for key in keylist])
+    array_2 = np.array([dict2[key] for key in keylist])
+
+    assert np.allclose(array_1, array_2)
+
+
+def test_bas():
+    ring = Ring()
+
+    h.finitialize(-65 * mV)
+    prun(50 * ms)
+    compare_dicts(
+        get_all_spikes(ring),
+        {
+            0: [10.925000000099914],
+            2: [37.40000000009994],
+            4: [],
+            1: [10.925000000099914, 37.42500000009995],
+            3: [],
+        },
+    )
+    prun(100 * ms, True)
```

Comment:
Note that SaveState is more or less equivalent to the checkpointing API for CoreNEURON (and BBSaveState is a bit more general than either because it is agnostic about the distribution of cells for restore.)